src/start_server.ts

```diff
@@ -182,6 +195,257 @@ function startWithPort(userOptions: ServerOptions) {
+    //
+    // While TypeScript 2.0 already supports it, tsc does not transpile async/await
+    // to ES5, which is scheduled for TypeScript 2.1. The advantages of async/await
+    // outweigh that of array destructuring, so go the verbose way for now...
+
+    const certData = await Promise.all([
+                                         fs.readFile(certPath)
+                                           .then((value: Buffer) => value.toString().trim()),
+                                         fs.readFile(keyPath)
+                                           .then((value: Buffer) => value.toString().trim())
+                                       ]);
+    const cert: string = certData[0];
+    const key: string = certData[1];
+    if (key && cert) {
+      certObj = {
+        cert: cert,
+        key: key,
+      };
+    }
+  }
+
+  if (!certObj) {
+    certObj = await createTLSCertificate();
+
+    if (keyPath && certPath) {
+      const _ = await Promise.all([
+                                    fs.writeFile(certPath, certObj.cert),
+                                    fs.writeFile(keyPath, certObj.key)
+                                  ]);
+    }
+  }
+
+  return certObj;
+}
+
+/**
+ * Asserts that Node version is valid for h2 protocol
+ * @param {ServerOptions} options
+ */
+function assertNodeVersion(options: ServerOptions) {
+  if (options.protocol === 'h2') {
+    const matches = /(\d+)\./.exec(process.version);
+    if (matches) {
+      const major = Number(matches[1]);
+      assert(major >= 5, 'h2 requires ALPN which is only supported in node.js >= 5.0');
+    }
+  }
+}
+
+/**
+ * Creates an HTTP(S) server
+ * @param app
+ * @param {ServerOptions} options
+ * @returns {Promise<http.Server>} Promise of server
+ */
+async function createServer(app: any, options: ServerOptions): Promise<http.Server> {
+  const opt: any = {
+    spdy: {protocols: [options.protocol]}
+  };
+
+  if (isHttps(options.protocol)) {
+    const keys = await getTLSCertificate(options.keyPath, options.certPath);
+    opt.key = keys.key;
+    opt.cert = keys.cert;
+  } else {
+    opt.spdy.plain = true;
+    opt.spdy.ssl = false;
+  }
+
+  return http.createServer(opt, app);
+}
+
+/**
+ * Starts an HTTP(S) server on a specific port
+ * @param {ServerOptions} userOptions
+ * @returns {Promise<http.Server>} Promise of server
+ */
+async function startWithPort(userOptions: ServerOptions): Promise<http.Server> {
+  const options = await applyDefaultOptions(userOptions);
+  const app = getApp(options);
 
-  return serverStartedPromise;
+  const server = await createServer(app, options);
+  server.listen(options.port, options.hostname, () => {
+    handleServerReady(options);
+  });
+
+  server.on('error', (err: any) => {
+    if (err.code === 'EADDRINUSE') {
+      console.error(portInUseMessage(options.port));
+    }
+    console.warn('rejecting with err', err);
+    throw new Error(err);
```

Comment:
This won't reject the startWithPort returned promise, but it would be nice if it did.